File: Review-07-Java-Toys/Example-3/src/main/java/edu/odu/cs/cs350/examples/Student.java
package edu.odu.cs.cs350.examples;

public class Student implements Cloneable {
    /**
     * Constant defining the default student name
     */
    public static final String DEFAULT_NAME = "John Q. Smith";

    private String name;

    /**
     * Create a Student with the default name
     * of John Q. Smith
     */
    public Student()
    {
        this.name = DEFAULT_NAME;
    }

    /**
     * Create a Student with a specified name
     *
     * @param n desired name
     */
    public Student(String n)
    {
        this.name = n;
    }

    /**
     * Get the Name
     */
    public String getName()
    {
        return name;
    }

    /**
     * Change the student name
     *
     * @param n desired name
     */
    public void setName(String n)
    {
        name = n;
    }

    /**
     * Compare 2 `Student`s based on name
     *
     * @param rhs the other (right-hand-side) student object
     */
    @Override
    public boolean equals(Object rhs)
    {
        if(!(rhs instanceof Student)) {
            return false;
        }

        return (this.name).equals(((Student)rhs).name);
    }

    /**
     * Return a hashcode
     */
    @Override
    public int hashCode()
    {
        return name.hashCode();
    }

    /**
     * Return a (deep) copy of this object.
     */
    @Override
    public Object clone()
    {
        return new Student(name);
    }

    /**
     *
     */
    @Override
    public String toString()
    {
        return name;
    }
}

File: Review-00-Foundational-Concepts/Classes-Composition-Python/examples/student.py
import copy

DEFAULT_NAME = "John Q. Smith"


class Student:
    def __init__(self, n: str = DEFAULT_NAME):
        self.__name = n

    @property
    def name(self):
        return self.__name

    @name.setter
    def name(self, n: str):
        self.__name = n

    def __eq__(self, rhs):
        """
        Compare 2 `Student`s based on name.
        """

        if not isinstance(rhs, Student):
            return False

        return self.__name == rhs.name

    def __hash__(self):
        return hash(self.__name)

    def __deepcopy__(self, memo):
        return Student(self.__name)

    def __str__(self):
        return self.__name

    def __repr__(self):
        return f'Student(n="{self.name}")'
